File: plantuml_generators/PlantUMLGenerator.cs
/**
 * PlantUML代码生成器 - C#版本
 * 支持生成各种类型的UML图表
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace PlantUMLGenerators
{
    public enum DiagramType
    {
        Activity,
        Sequence,
        Class,
        UseCase,
        Component,
        State,
        Deployment
    }

    public class Note
    {
        public string Content { get; set; }
        public string Position { get; set; }

        public Note(string content, string position = "bottom")
        {
            Content = content;
            Position = position;
        }
    }

    public class Activity
    {
        public string Type { get; set; }
        public string Name { get; set; }
        public string Note { get; set; }
        public string NotePosition { get; set; }

        public Activity(string name, string note = "", string notePosition = "right")
        {
            Type = "activity";
            Name = name;
            Note = note;
            NotePosition = notePosition;
        }
    }

    public class Partition
    {
        public string Name { get; set; }
        public List<string> Activities { get; set; }
        public string Color { get; set; }

        public Partition(string name, List<string> activities, string color = "")
        {
            Name = name;
            Activities = activities;
            Color = color;
        }
    }

    public class Decision
    {
        public string Question { get; set; }
        public string YesPath { get; set; }
        public string NoPath { get; set; }
        public string YesNote { get; set; }
        public string NoNote { get; set; }

        public Decision(string question, string yesPath, string noPath, 
                       string yesNote = "", string noNote = "")
        {
            Question = question;
            YesPath = yesPath;
            NoPath = noPath;
            YesNote = yesNote;
            NoNote = noNote;
        }
    }

    public abstract class PlantUMLGenerator
    {
        protected string theme;
        protected string title;
        protected List<Note> notes;

        protected PlantUMLGenerator()
        {
            theme = "plain";
            title = "";
            notes = new List<Note>();
        }

        public virtual PlantUMLGenerator SetTheme(string theme = "plain")
        {
            this.theme = theme;
            return this;
        }

        public virtual PlantUMLGenerator SetTitle(string title)
        {
            this.title = title;
            return this;
        }

        public virtual PlantUMLGenerator AddNote(string content, string position = "bottom")
        {
            notes.Add(new Note(content, position));
            return this;
        }

        public abstract string GenerateCode();
    }

    public class ActivityDiagramGenerator : PlantUMLGenerator
    {
        private List<Activity> activities;
        private List<Partition> partitions;
        private List<Decision> decisions;
        private List<List<string>> parallelBlocks;

        public ActivityDiagramGenerator()
        {
            activities = new List<Activity>();
            partitions = new List<Partition>();
            decisions = new List<Decision>();
            parallelBlocks = new List<List<string>>();
        }

        public ActivityDiagramGenerator AddActivity(string name, string note = "", string notePosition = "right")
        {
            activities.Add(new Activity(name, note, notePosition));
            return this;
        }

        public ActivityDiagramGenerator AddPartition(string name, List<string> activities, string color = "")
        {
            partitions.Add(new Partition(name, activities, color));
            return this;
        }

        public ActivityDiagramGenerator AddDecision(string question, string yesPath, string noPath,
                                                   string yesNote = "", string noNote = "")
        {
            decisions.Add(new Decision(question, yesPath, noPath, yesNote, noNote));
            return this;
        }

        public ActivityDiagramGenerator AddParallelBlock(List<string> activities)
        {
            parallelBlocks.Add(new List<string>(activities));
            return this;
        }

        public override string GenerateCode()
        {
            var code = new List<string> { "@startuml" };

            if (!string.IsNullOrEmpty(theme))
            {
                code.Add($"!theme {theme}");
            }

            if (!string.IsNullOrEmpty(title))
            {
                code.Add($"title {title}");
            }

            // 添加样式定义
            code.AddRange(new[]
            {
                "",
                "' 样式定义",
                "skinparam activity {",
                "    BackgroundColor lightblue",
                "    BorderColor black",
                "    FontSize 12",
                "}",
                "",
                "skinparam activityDiamond {",
                "    BackgroundColor lightyellow",
                "    BorderColor black",
                "}",
                "",
                "skinparam note {",
                "    BackgroundColor lightgreen",
                "    BorderColor darkgreen",
                "}",
                "",
                "start",
                ""
            });

            // 生成分区和活动
            foreach (var partition in partitions)
            {
                var colorAttr = string.IsNullOrEmpty(partition.Color) ? "" : $" #{partition.Color}";
                code.Add($"partition \"{partition.Name}\"{colorAttr} {{");

                foreach (var activity in partition.Activities)
                {
                    code.Add($"    :{activity};");
                }

                code.Add("}");
                code.Add("");
            }

            // 添加单独的活动
            foreach (var activity in activities)
            {
                code.Add($":{activity.Name};");
                if (!string.IsNullOrEmpty(activity.Note))
                {
                    code.Add($"note {activity.NotePosition}: {activity.Note}");
                }
                code.Add("");
            }

            // 添加决策点
            foreach (var decision in decisions)
            {
                code.Add($"if ({decision.Question}) then (是)");
                code.Add($"    :{decision.YesPath};");
                if (!string.IsNullOrEmpty(decision.YesNote))
                {
                    code.Add($"    note right: {decision.YesNote}");
                }
                code.Add("else (否)");
                code.Add($"    :{decision.NoPath};");
                if (!string.IsNullOrEmpty(decision.NoNote))
                {
                    code.Add($"    note right: {decision.NoNote}");
                }
                code.Add("endif");
                code.Add("");
            }

            // 添加并行块
            foreach (var parallel in parallelBlocks)
            {
                code.Add("fork");
                for (int i = 0; i < parallel.Count; i++)
                {
                    if (i > 0)
                    {
                        code.Add("fork again");
                    }
                    code.Add($"    :{parallel[i]};");
                }
                code.Add("end fork");
                code.Add("");
            }

            code.Add("stop");

            // 添加注释
            foreach (var note in notes)
            {
                code.Add($"note {note.Position}");
                code.Add($"    {note.Content}");
                code.Add("end note");
            }

            code.Add("@enduml");

            return string.Join("\n", code);
        }
    }

    public class PlantUMLFactory
    {
        public static PlantUMLGenerator CreateGenerator(DiagramType diagramType)
        {
            switch (diagramType)
            {
                case DiagramType.Activity:
                    return new ActivityDiagramGenerator();
                default:
                    return new ActivityDiagramGenerator();
            }
        }
    }

    public class Program
    {
        public static string CreateBuildingAIWorkflow()
        {
            var generator = new ActivityDiagramGenerator();

            generator.SetTitle("建筑AI辅助策划与设计系统工作流程");

            // 添加分区和活动
            generator.AddPartition("数据输入阶段", new List<string>
            {
                "收集既往建筑项目策划书",
                "收集项目任务书",
                "获取国家建设标准",
                "获取地方建设标准"
            });

            generator.AddPartition("AI分析处理阶段", new List<string>
            {
                "建筑大语言模型加载",
                "多源数据融合分析",
                "启动对话式交互"
            });

            generator.AddPartition("智能生成阶段", new List<string>
            {
                "快速生成策划书草案",
                "生成设计任务书",
                "识别关键性能指标",
                "明确项目目标和需求"
            });

            generator.AddPartition("输出优化阶段", new List<string>
            {
                "为设计师提供清晰方向",
                "识别技术挑战点",
                "生成针对性解决方案"
            });

            generator.AddPartition("价值实现阶段", new List<string>
            {
                "降低认知负荷",
                "提高工作效率",
                "优化设计质量"
            });

            generator.AddNote(
                "系统通过AI技术整合多源建筑信息，实现智能化的项目策划与设计辅助",
                "top"
            );

            return generator.GenerateCode();
        }

        public static void Main(string[] args)
        {
            Console.WriteLine("🚀 PlantUML生成器 - C#版本");
            Console.WriteLine(new string('=', 50));

            try
            {
                // 创建输出目录
                string outputDir = "/workspace/plantuml_generators/csharp_output";
                Directory.CreateDirectory(outputDir);

                // 生成建筑AI工作流程图
                string workflowCode = CreateBuildingAIWorkflow();
                string workflowFile = Path.Combine(outputDir, "building_ai_workflow.puml");
                File.WriteAllText(workflowFile, workflowCode, Encoding.UTF8);

                Console.WriteLine("✅ 已生成以下文件：");
                Console.WriteLine($"   📄 工作流程图: {workflowFile}");
                Console.WriteLine($"\n🎯 所有文件已保存到: {outputDir}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"错误: {e.Message}");
            }
        }
    }
}

File: plantuml_generators/PlantUMLGenerator.java
/**
 * PlantUML代码生成器 - Java版本
 * 支持生成各种类型的UML图表
 */

import java.util.*;
import java.io.*;
import java.nio.file.*;

public class PlantUMLGenerator {
    protected String theme;
    protected String title;
    protected List<Note> notes;
    
    public PlantUMLGenerator() {
        this.theme = "plain";
        this.title = "";
        this.notes = new ArrayList<>();
    }
    
    public PlantUMLGenerator setTheme(String theme) {
        this.theme = theme;
        return this;
    }
    
    public PlantUMLGenerator setTitle(String title) {
        this.title = title;
        return this;
    }
    
    public PlantUMLGenerator addNote(String content, String position) {
        this.notes.add(new Note(content, position));
        return this;
    }
    
    public String generateCode() {
        return "@startuml\n@enduml";
    }
    
    // 内部类定义
    static class Note {
        String content;
        String position;
        
        Note(String content, String position) {
            this.content = content;
            this.position = position;
        }
    }
    
    static class Activity {
        String name;
        String note;
        String notePosition;
        
        Activity(String name, String note, String notePosition) {
            this.name = name;
            this.note = note;
            this.notePosition = notePosition;
        }
    }
    
    static class Partition {
        String name;
        List<String> activities;
        String color;
        
        Partition(String name, List<String> activities, String color) {
            this.name = name;
            this.activities = activities;
            this.color = color;
        }
    }
    
    static class Decision {
        String question;
        String yesPath;
        String noPath;
        String yesNote;
        String noNote;
        
        Decision(String question, String yesPath, String noPath, String yesNote, String noNote) {
            this.question = question;
            this.yesPath = yesPath;
            this.noPath = noPath;
            this.yesNote = yesNote;
            this.noNote = noNote;
        }
    }
}

class ActivityDiagramGenerator extends PlantUMLGenerator {
    private List<Activity> activities;
    private List<Partition> partitions;
    private List<Decision> decisions;
    private List<List<String>> parallelBlocks;
    
    public ActivityDiagramGenerator() {
        super();
        this.activities = new ArrayList<>();
        this.partitions = new ArrayList<>();
        this.decisions = new ArrayList<>();
        this.parallelBlocks = new ArrayList<>();
    }
    
    public ActivityDiagramGenerator addActivity(String name, String note, String notePosition) {
        this.activities.add(new Activity(name, note, notePosition));
        return this;
    }
    
    public ActivityDiagramGenerator addActivity(String name) {
        return addActivity(name, "", "right");
    }
    
    public ActivityDiagramGenerator addPartition(String name, List<String> activities, String color) {
        this.partitions.add(new Partition(name, activities, color));
        return this;
    }
    
    public ActivityDiagramGenerator addPartition(String name, List<String> activities) {
        return addPartition(name, activities, "");
    }
    
    public ActivityDiagramGenerator addDecision(String question, String yesPath, String noPath, 
                                              String yesNote, String noNote) {
        this.decisions.add(new Decision(question, yesPath, noPath, yesNote, noNote));
        return this;
    }
    
    public ActivityDiagramGenerator addDecision(String question, String yesPath, String noPath) {
        return addDecision(question, yesPath, noPath, "", "");
    }
    
    public ActivityDiagramGenerator addParallelBlock(List<String> activities) {
        this.parallelBlocks.add(new ArrayList<>(activities));
        return this;
    }
    
    @Override
    public String generateCode() {
        StringBuilder code = new StringBuilder();
        
        code.append("@startuml\n");
        
        if (theme != null && !theme.isEmpty()) {
            code.append("!theme ").append(theme).append("\n");
        }
        
        if (title != null && !title.isEmpty()) {
            code.append("title ").append(title).append("\n");
        }
        
        // 添加样式定义
        code.append("\n' 样式定义\n");
        code.append("skinparam activity {\n");
        code.append("    BackgroundColor lightblue\n");
        code.append("    BorderColor black\n");
        code.append("    FontSize 12\n");
        code.append("}\n\n");
        
        code.append("skinparam activityDiamond {\n");
        code.append("    BackgroundColor lightyellow\n");
        code.append("    BorderColor black\n");
        code.append("}\n\n");
        
        code.append("skinparam note {\n");
        code.append("    BackgroundColor lightgreen\n");
        code.append("    BorderColor darkgreen\n");
        code.append("}\n\n");
        
        code.append("start\n\n");
        
        // 生成分区和活动
        for (Partition partition : partitions) {
            String colorAttr = partition.color.isEmpty() ? "" : " #" + partition.color;
            code.append("partition \"").append(partition.name).append("\"").append(colorAttr).append(" {\n");
            
            for (String activity : partition.activities) {
                code.append("    :").append(activity).append(";\n");
            }
            
            code.append("}\n\n");
        }
        
        // 添加单独的活动
        for (Activity activity : activities) {
            code.append(":").append(activity.name).append(";\n");
            if (!activity.note.isEmpty()) {
                code.append("note ").append(activity.notePosition).append(": ").append(activity.note).append("\n");
            }
            code.append("\n");
        }
        
        // 添加决策点
        for (Decision decision : decisions) {
            code.append("if (").append(decision.question).append(") then (是)\n");
            code.append("    :").append(decision.yesPath).append(";\n");
            if (!decision.yesNote.isEmpty()) {
                code.append("    note right: ").append(decision.yesNote).append("\n");
            }
            code.append("else (否)\n");
            code.append("    :").append(decision.noPath).append(";\n");
            if (!decision.noNote.isEmpty()) {
                code.append("    note right: ").append(decision.noNote).append("\n");
            }
            code.append("endif\n\n");
        }
        
        // 添加并行块
        for (List<String> parallel : parallelBlocks) {
            code.append("fork\n");
            for (int i = 0; i < parallel.size(); i++) {
                if (i > 0) {
                    code.append("fork again\n");
                }
                code.append("    :").append(parallel.get(i)).append(";\n");
            }
            code.append("end fork\n\n");
        }
        
        code.append("stop\n");
        
        // 添加注释
        for (Note note : notes) {
            code.append("note ").append(note.position).append("\n");
            code.append("    ").append(note.content).append("\n");
            code.append("end note\n");
        }
        
        code.append("@enduml");
        
        return code.toString();
    }
}

class SequenceDiagramGenerator extends PlantUMLGenerator {
    private List<Participant> participants;
    private List<Message> messages;
    
    public SequenceDiagramGenerator() {
        super();
        this.participants = new ArrayList<>();
        this.messages = new ArrayList<>();
    }
    
    public SequenceDiagramGenerator addParticipant(String name, String alias, String participantType) {
        this.participants.add(new Participant(name, alias, participantType));
        return this;
    }
    
    public SequenceDiagramGenerator addParticipant(String name) {
        return addParticipant(name, name, "participant");
    }
    
    public SequenceDiagramGenerator addMessage(String from, String to, String message, String arrowType) {
        this.messages.add(new Message(from, to, message, arrowType));
        return this;
    }
    
    public SequenceDiagramGenerator addMessage(String from, String to, String message) {
        return addMessage(from, to, message, "->");
    }
    
    @Override
    public String generateCode() {
        StringBuilder code = new StringBuilder();
        
        code.append("@startuml\n");
        
        if (theme != null && !theme.isEmpty()) {
            code.append("!theme ").append(theme).append("\n");
        }
        
        if (title != null && !title.isEmpty()) {
            code.append("title ").append(title).append("\n");
        }
        
        code.append("\n");
        
        // 添加参与者
        for (Participant participant : participants) {
            if (!participant.alias.equals(participant.name)) {
                code.append(participant.type).append(" \"").append(participant.name)
                    .append("\" as ").append(participant.alias).append("\n");
            } else {
                code.append(participant.type).append(" ").append(participant.name).append("\n");
            }
        }
        
        code.append("\n");
        
        // 添加消息
        for (Message msg : messages) {
            code.append(msg.from).append(" ").append(msg.arrow).append(" ")
                .append(msg.to).append(": ").append(msg.message).append("\n");
        }
        
        code.append("@enduml");
        
        return code.toString();
    }
    
    static class Participant {
        String name;
        String alias;
        String type;
        
        Participant(String name, String alias, String type) {
            this.name = name;
            this.alias = alias;
            this.type = type;
        }
    }
    
    static class Message {
        String from;
        String to;
        String message;
        String arrow;
        
        Message(String from, String to, String message, String arrow) {
            this.from = from;
            this.to = to;
            this.message = message;
            this.arrow = arrow;
        }
    }
}

// 主类和示例用法
public class PlantUMLGeneratorMain {
    
    public static String createBuildingAIWorkflow() {
        ActivityDiagramGenerator generator = new ActivityDiagramGenerator();
        
        generator.setTitle("建筑AI辅助策划与设计系统工作流程");
        
        // 添加分区和活动
        generator.addPartition("数据输入阶段", Arrays.asList(
            "收集既往建筑项目策划书",
            "收集项目任务书",
            "获取国家建设标准",
            "获取地方建设标准"
        ));
        
        generator.addPartition("AI分析处理阶段", Arrays.asList(
            "建筑大语言模型加载",
            "多源数据融合分析",
            "启动对话式交互"
        ));
        
        generator.addPartition("智能生成阶段", Arrays.asList(
            "快速生成策划书草案",
            "生成设计任务书",
            "识别关键性能指标",
            "明确项目目标和需求"
        ));
        
        generator.addPartition("输出优化阶段", Arrays.asList(
            "为设计师提供清晰方向",
            "识别技术挑战点",
            "生成针对性解决方案"
        ));
        
        generator.addPartition("价值实现阶段", Arrays.asList(
            "降低认知负荷",
            "提高工作效率",
            "优化设计质量"
        ));
        
        generator.addNote(
            "系统通过AI技术整合多源建筑信息，实现智能化的项目策划与设计辅助",
            "top"
        );
        
        return generator.generateCode();
    }
    
    public static void saveToFile(String content, String filePath) throws IOException {
        Files.write(Paths.get(filePath), content.getBytes("UTF-8"));
    }
    
    public static void main(String[] args) {
        System.out.println("🚀 PlantUML生成器 - Java版本");
        System.out.println("=".repeat(50));
        
        try {
            // 创建输出目录
            String outputDir = "/workspace/plantuml_generators/java_output";
            Files.createDirectories(Paths.get(outputDir));
            
            // 生成建筑AI工作流程图
            String workflowCode = createBuildingAIWorkflow();
            String workflowFile = outputDir + "/building_ai_workflow.puml";
            saveToFile(workflowCode, workflowFile);
            
            // 生成序列图示例
            SequenceDiagramGenerator seqGen = new SequenceDiagramGenerator();
            seqGen.setTitle("建筑AI系统交互序列图")
                  .addParticipant("用户", "User", "actor")
                  .addParticipant("AI系统", "AI", "participant")
                  .addParticipant("数据库", "DB", "database")
                  .addMessage("User", "AI", "上传项目信息")
                  .addMessage("AI", "DB", "查询历史数据")
                  .addMessage("AI", "User", "生成策划书");
            
            String sequenceCode = seqGen.generateCode();
            String sequenceFile = outputDir + "/sequence_diagram.puml";
            saveToFile(sequenceCode, sequenceFile);
            
            System.out.println("✅ 已生成以下文件：");
            System.out.println("   📄 工作流程图: " + workflowFile);
            System.out.println("   📄 序列图示例: " + sequenceFile);
            System.out.println("\n🎯 所有文件已保存到: " + outputDir);
            
        } catch (IOException e) {
            System.err.println("文件操作错误: " + e.getMessage());
        }
    }
}